### ISSUE TITLE
Missing data on failing output (multi output)

### ISSUE THREAD

**Issue by bshifter** [fluent/fluent-bit#2850]
## Bug Report

**Describe the bug**
We’ve found a possible issue about multiple outputs. We aren’t sure whether it’s an intended decision or a bug, or just a simple configurable behaviour, since we haven’t found any information in the docs about. 
When multiple outputs are defined it seems that one healthy output is enough to clear chunks from the buffer. Because of that the non-healthy output won’t receive the chunks generated during the outage. After it becomes healthy again both of them receive the new chunks as expected. The result is that we have a gap in the data of the failing output.

**To Reproduce**
- Steps to reproduce the problem:

### Config
```[SERVICE]
    Flush        1
    Grace        5
    Daemon       Off
    Log_Level    info
    Parsers_File parsers.conf
    Coro_Stack_Size    24576
    HTTP_Server  On
    HTTP_Listen  0.0.0.0
    HTTP_Port    2020
    storage.backlog.mem_limit  256KB
    storage.path  /buffers

[INPUT]
    Name         tail
    DB  /tail-db/tail-containers-state.db
    Mem_Buf_Limit  256KB
    Parser  cri
    Path  /var/log/containers/*.log
    Refresh_Interval  5
    Skip_Long_Lines  On
    Tag  kubernetes.*
    storage.type  filesystem

[FILTER]
    Name        kubernetes
    Kube_CA_File  /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    Kube_Tag_Prefix  kubernetes.var.log.containers
    Kube_Token_File  /var/run/secrets/kubernetes.io/serviceaccount/token
    Kube_URL  https://kubernetes.default.svc:443
    Match  kubernetes.*
    Merge_Log  On

[OUTPUT]
    Name          forward
    Match         *
    Host          one-eye-fluentd.default.svc
    Port          24240

    Retry_Limit  False

[OUTPUT]
    Name loki
    Match *
    Host one-eye-loki.default.svc
    Port 3100
    Labels job=fluentbit
    Auto_kubernetes_labels off
    Label_keys $kubernetes['host'], $kubernetes['pod_name'], $kubernetes['container_name']
    Retry_Limit False
```
The configuration works as expected:
![image2](https://user-images.githubusercontent.com/39086512/101763148-ad534b80-3ade-11eb-880a-05485580399b.png)
	
### Make loki unavailable
```
kubectl patch sts one-eye-loki --type merge -p "
spec:
 replicas: 0"
```
```
~> k logs -l app.kubernetes.io/name=fluentbit -f

[2020/12/08 10:06:13] [ warn] [engine] failed to flush chunk '1-1607421972.613273700.flb', retry in 9 seconds: task_id=0, input=tail.0 > output=loki.1
[2020/12/08 10:06:22] [error] [output:loki:loki.1] one-eye-loki.default.svc:3100, HTTP status=500
at least 1 live replicas required, could only find 0

[2020/12/08 10:06:22] [ warn] [engine] failed to flush chunk '1-1607421972.613273700.flb', retry in 7 seconds: task_id=0, input=tail.0 > output=loki.1
[2020/12/08 10:06:30] [error] [io] TCP connection failed: one-eye-loki.default.svc:3100 (Connection refused)
[2020/12/08 10:06:30] [error] [output:loki:loki.1] no upstream connections available
[2020/12/08 10:06:30] [ warn] [engine] failed to flush chunk '1-1607421972.613273700.flb', retry in 976 seconds: task_id=0, input=tail.0 > output=loki.1
[2020/12/08 10:06:43] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=2495306 watch_fd=20
```

### Checking the buffers in fluent-bit pod shows some interesting results
```
/buffers/tail.0 # date; ls -l
Tue Dec  8 10:10:40 UTC 2020
total 24
-rw-------    1 root     root          8316 Dec  8 10:06 1-1607421972.613273700.flb
-rw-------    1 root     root          4096 Dec  8 10:10 1-1607422240.308303600.flb
-rw-------    1 root     root          4096 Dec  8 10:10 1-1607422240.348164900.flb
-rw-------    1 root     root          4096 Dec  8 10:10 1-1607422240.389524200.flb
/buffers/tail.0 # date; ls -l
Tue Dec  8 10:10:49 UTC 2020
total 120
-rw-------    1 root     root          8316 Dec  8 10:06 1-1607421972.613273700.flb
-rw-------    1 root     root         36864 Dec  8 10:10 1-1607422249.294289200.flb
-rw-------    1 root     root         36864 Dec  8 10:10 1-1607422249.296574600.flb
-rw-------    1 root     root         36864 Dec  8 10:10 1-1607422249.324124200.flb
/buffers/tail.0 # date; ls -l
Tue Dec  8 10:10:58 UTC 2020
total 24
-rw-------    1 root     root          8316 Dec  8 10:06 1-1607421972.613273700.flb
-rw-------    1 root     root          4096 Dec  8 10:10 1-1607422258.306430500.flb
-rw-------    1 root     root          4096 Dec  8 10:10 1-1607422258.329320600.flb
-rw-------    1 root     root          4096 Dec  8 10:10 1-1607422258.344480100.flb
```

Only a single chunk is retained, however its size does not change. All the other chunks are disappearing periodically. Seemingly they get flushed.

### Restore loki
```
~> kubectl patch sts one-eye-loki --type merge -p "
spec:
  replicas: 1"
```
```
[2020/12/08 10:06:13] [ warn] [engine] failed to flush chunk '1-1607421972.613273700.flb', retry in 9 seconds: task_id=0, input=tail.0 > output=loki.1
[2020/12/08 10:06:22] [error] [output:loki:loki.1] one-eye-loki.default.svc:3100, HTTP status=500
at least 1 live replicas required, could only find 0

[2020/12/08 10:06:22] [ warn] [engine] failed to flush chunk '1-1607421972.613273700.flb', retry in 7 seconds: task_id=0, input=tail.0 > output=loki.1
[2020/12/08 10:06:30] [error] [io] TCP connection failed: one-eye-loki.default.svc:3100 (Connection refused)
[2020/12/08 10:06:30] [error] [output:loki:loki.1] no upstream connections available
[2020/12/08 10:06:30] [ warn] [engine] failed to flush chunk '1-1607421972.613273700.flb', retry in 976 seconds: task_id=0, input=tail.0 > output=loki.1
[2020/12/08 10:06:43] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=2495306 watch_fd=20
[2020/12/08 10:13:54] [ info] [input:tail:tail.0] inotify_fs_add(): inode=2632326 watch_fd=45 name=/var/log/containers/one-eye-loki-0_default_loki-0a865079f3a9848a183e13b81db88672937a9d9e2a476f0f0ee3767d129c309b.log
[2020/12/08 10:22:46] [ info] [output:loki:loki.1] one-eye-loki.default.svc:3100, HTTP status=204
[2020/12/08 10:22:46] [ info] [engine] flush chunk '1-1607421972.613273700.flb' succeeded at retry 3: task_id=4, input=tail.0 > output=loki.1
[2020/12/08 10:22:47] [ info] [output:loki:loki.1] one-eye-loki.default.svc:3100, HTTP status=204
[2020/12/08 10:22:48] [ info] [output:loki:loki.1] one-eye-loki.default.svc:3100, HTTP status=204
```
![image1](https://user-images.githubusercontent.com/39086512/101763149-ae847880-3ade-11eb-9abe-a4209cfa7a23.png)

There are no logs for the particular time range in loki. It seems fluent-bit does not wait for the loki output to be able to flush buffer chunks again because the forward output successfully finishes the flushing.

Furthermore, without the forward output configuration (single loki output), buffering chunks looks completely different. All the data gets buffered on the fluent-bit pod while loki returns, and fluent-bit is able to flush them.

**Expected behavior**
Failing output receives all the data after the outage.

**Your Environment**

* Version used: fluent-bit 1.6.6 (1.6.6-debug) official fluent-bit containers
* Configuration: see above

**Comment by JeffLuoo** [comment]
Thanks for reporting this issue. I will be looking into it.

**Comment by JeffLuoo** [comment]
@bshifter I came back and reviewed this case again. Here is the link to the [design doc](https://github.com/fluent/fluent-bit/issues/2408) of the dropping oldest chunk when the limit is reached. From my design, I intended to keep the latest chunk when the limit is reached. And one of the conditions to remove a chunk from the disk is that the chunk is not being forwarded to the destination. So this might result in the case you mentioned on the issue.

@erain Hi Yu, WDYT on this issue?
